Mathlib/Topology/UniformSpace/Dini.lean

```diff
@@ -0,0 +1,97 @@
+/-
+Copyright (c) 2024 Jireh Loreaux. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+import Mathlib.Topology.Algebra.UniformGroup.Basic
+import Mathlib.Topology.ContinuousMap.Ordered
+import Mathlib.Topology.Order.MonotoneConvergence
+import Mathlib.Topology.UniformSpace.CompactConvergence
+
+/-! # Dini's Theorem
+
+This file proves Dini's theorem, which states that if `F n` is a monotone increasing sequence of
+continuous real-valued functions on a compact set `s` converging pointwise to a continuous function
+`f`, then `F n` converges uniformly to `f`.
+
+We generalize the codomain from `ℝ` to a linearly ordered commutative group `G` with the order
+topology which is also order closed. Therefore it applies also when the codomain is `ℤ`, `ℚ`, or
+`ℝ≥0ˣ`.
+-/
+
+open Filter Topology
+
+variable {ι α G : Type*} [SemilatticeSup ι]
+    [Nonempty ι] [TopologicalSpace α] [LinearOrderedCommGroup G] [UniformSpace G]
+    [UniformGroup G] [OrderTopology G] [OrderClosedTopology G] [Nontrivial G]
+
+namespace UniformGroup
+
+variable {F : ι → α → G} {f : α → G}
+
+/-- **Dini's theorem** If `F n` is a monotone increasing collection of continuous functions on a
+compact space converging pointwise to a continuous `f`, then `F n` converges uniformly to `f`. -/
+@[to_additive "**Dini's theorem** If `F n` is a monotone increasing collection of continuous
+functions on a compact space converging pointwise to a continuous `f`, then `F n` converges
+uniformly to `f`."]
+lemma tendstoUniformly_of_forall_tendsto [CompactSpace α] (hF_cont : ∀ i, Continuous (F i))
```

Comment:
I have a bunch of variants of the statement to suggest, not all of which are necessarily interesting right now.
- I think it would be nice to first state the [TendstoLocallyUniformly](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/UniformConvergence.html#TendstoLocallyUniformly) (or maybe even [TendstoUniformlyOnFilter](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/UniformConvergence.html#TendstoUniformlyOnFilter)) statement, which completely gets rid of the compactness assumption. Then this result comes for free thanks to [tendstoLocallyUniformly_iff_tendstoUniformly_of_compactSpace](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/UniformSpace/UniformConvergence.html#tendstoLocallyUniformly_iff_tendstoUniformly_of_compactSpace)
- I think you can just immediately replace `SemilatticeSup G` by [IsDirected](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Directed.html#IsDirected)
- In the same vein, one could add a version of the statement about infimas of directed families, instead of limits. I'm not sure this one should be added right now, but I mention it in case you think it should.

Reply:
+ I don't think we can get rid of compactness, but maybe I'm wrong.
+ I switched to `IsDirected`
+ I'll probably add this when I have time.

Reply:
(I forgot to answer the last point : indeed this is what I had in mind you need some hypothesis like that)